test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -0,0 +1,710 @@
+ * }
+ *
+ * <p>
+ * <strong>Details:</strong>
+ * <p>
+ * A {@link Template} can have zero or more arguments. A template can be created with {@code make} methods like
+ * {@link Template#make(String, Function)}. For each number of arguments there is an implementation
+ * (e.g. {@link Template.TwoArgs} for two arguments). This allows the use of Generics for the
+ * Template argument types which enables type checking of the Template arguments.
+ *
+ * <p>
+ * A {@link Template} can be rendered to a {@link String} (e.g. {@link Template.ZeroArgs#render()}).
+ * Alternatively, we can generate a {@link Token} (more specifically, a {@link TemplateToken}) with {@code asToken()}
+ * (e.g. {@link Template.ZeroArgs#asToken()}), and use the {@link Token} inside another {@link Template#body}.
+ *
+ * <p>
+ * Ideally, we would have used {@link String} Templates to inject these Template arguments into the strings.
+ * But since {@link String} Templates are not (yet) available, the Templates provide <strong>hashtag replacements</strong>
+ * in the {@link String}s: the Template argument names are captured, and the argument values automatically replace any
+ * {@code "#name"} in the {@link String}s. See the different overloads of {@link #make} for examples. Additional hashtag
+ * replacements can be defined with {@link #let}.
+ *
+ * <p>
+ * When using nested Templates, there can be collisions with identifiers (e.g. variable names and method names).
+ * For this, Templates provide <strong>dollar replacements</strong>, which automatically rename any
+ * {@code "$name"} in the {@link String} with a {@code "name_ID"}, where the {@code "ID"} is unique for every use of
+ * a Template. The dollar replacement can also be captured with {@link #$}, and passed to nested
+ * Templates, which allows sharing of these identifier names between Templates.
+ *
+ * <p>
+ * A {@link TemplateToken} can not just be used in {@link Template#body}, but it can also be
+ * {@link Hook#insert}ed to where a {@link Hook} was {@link Hook#set} earlier (in some outer scope of the code).
+ * For example, while generating code in a method, one can reach out to the scope of the class, and insert a
+ * new field, or define a utility method.
+ *
+ * <p>
+ * A {@link TemplateBinding} allows the recursive use of Templates. With the indirection of such a binding,
+ * a Template can reference itself. To ensure the termination of recursion, the templates are rendered
+ * with a certain amount of {@link #fuel} (default: 100, see {@link #DEFAULT_FUEL}), which is decreased at each
+ * Template nesting by a certain amount (default: 10, see {@link #DEFAULT_FUEL_COST}). The default fuel for a
+ * template can be changed when we {@code render()} it (e.g. {@link ZeroArgs#render(float)}) and the default
+ * fuel cost with {@link #setFuelCost}) when defining the {@link #body(Object...)}. Recursive templates are
+ * supposed to terminate once the {@link #fuel} is depleted (i.e. reaches zero).
```

Comment:
This sentence is a bit vague, maybe you could state explicitly who is responsible for ensuring termination (e.g. "Once the {@link #fuel} is depleted (i.e. reaches zero), the writer of a recursive template should ensure ...").

Reply:
I updated the section a little. I hope it is better now?

Reply:
Yes, thanks! For better readability, I suggest merging the paragraph ending with "With the indirection of such a binding, a Template can reference itself." and the paragraph starting with "The writer of recursive {@link Template}s must ensure that this recursion terminates" into a single paragraph.